index.html

```diff
@@ -0,0 +1,396 @@
+      } else if (ansiCode === "\x1b[?25l") {
+        this._cursorIsShowing = false;
+      } else if (ansiCode.endsWith("H")) {
+        const [line, column] = ansiCode.slice(2, -1).split(';').map(x => +x);
+        this.moveCursor(column-1, line-1);
+      } else if (ansiCode.startsWith("\x1b[1;") && ansiCode.endsWith("m") && COLORS[ansiCode.slice(4, -1)]) {
+        const colorInfo = COLORS[ansiCode.slice(4, -1)];
+        if (colorInfo.fg) this.fgColor = colorInfo.fg;
+        if (colorInfo.bg) this.bgColor = colorInfo.bg;
+      } else {
+        console.warn("Unknown ANSI escape sequence: " + ansiCode);
+      }
+    }
+
+    addTextWithEscapeSequences(text) {
+      const chunks = text.match(/\x1b\[[^A-Za-z]+[A-Za-z]|\r|\n|[^\x1b\r\n]+/g);
+      for (const chunk of chunks) {
+        if (chunk === "\r") {
+          this.moveCursorToStartOfLine();
+        } else if (chunk === "\n") {
+          this.moveCursorDown();
+        } else if (chunk.startsWith("\x1b")) {
+          this._handleAnsiCode(chunk);
+        } else {
+          this.addText(chunk);
+        }
+      }
+
+      console.log(`There are now ${this._pre.children.length} spans`);
+      this._updateCursor();
+    }
+  };
+
+  const ws = new WebSocket(`ws://${window.location.hostname}:54321`);
+
+  document.onkeydown = event => {
+    if (event.ctrlKey || event.altKey || ws.readyState !== WebSocket.OPEN) {
+      return;
+    }
+
+    let textToSend;
+    if (event.key.length === 1) {
+      textToSend = event.key;
+    } else if (event.key === "ArrowUp") {
+      textToSend = "\x1b[A";
+    } else if (event.key === "ArrowDown") {
+      textToSend = "\x1b[B";
+    } else if (event.key === "ArrowRight") {
+      textToSend = "\x1b[C";
+    } else if (event.key === "ArrowLeft") {
+      textToSend = "\x1b[D";
+    } else if (event.key === "Backspace") {
+      textToSend = "\x7f";
+    } else if (event.key === "Enter") {
+      textToSend = "\r";
+    } else {
+      console.log("Unrecognized key:", event.key);
+      return;
+    }
+
+    const utf8 = new TextEncoder().encode(textToSend);
+    ws.send(utf8);
+
+    event.preventDefault();
+  };
+
+  // For some reason, the .text() method on Blob is asynchronous.
+  // We need to make sure they run in the correct order, and not at the same time.
+  //
+  // UTF-8 characters, ANSI escape sequences etc are never received in two parts.
+  // This is because of how websockets work, and isn't true for TCP.
+
+  let receivedBlobs = [];
+  let handleBlobsRunning = false;
+
+  async function handleBlobs() {
+    while (receivedBlobs.length !== 0) {
+      const blobs = receivedBlobs;
+      receivedBlobs = [];
+      const texts = await Promise.all(blobs.map(b => b.text()));
+
+      if (ws.readyState === WebSocket.OPEN) {
+        terminal.addTextWithEscapeSequences(texts.join(""));
+      }
+    }
+    handleBlobsRunning = false;
+  }
+
+  ws.onmessage = (msg) => {
+    receivedBlobs.push(msg.data);
```

Comment:
Might be good to call `.text()` here, so that it has as much time to run as possible